### PR TITLE
New vverbosity level used to show spread output on real time

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This option is used to show the stdout and stderr in the console when running spread tasks
The option -vv which showd debug info has been moved to -vvv
A new section in the docs has been added to explain how the different levels of verbosity work.

Test it by running:
spread -vvgoogle:ubuntu-16.04-64:tasks/....

### CODE REVIEW

**Comment by niemeyer** [comment]
This is not just showing the output.. it is losing the output by sending it straight to the terminal, changing the behavior of Spread. Follow up logic will work differently claiming that it got an empty buffer even when it didn't. Logged messages will be wrong.

You can already get the output of commands by using "spread -vv", which enables the spread.Debug flag and debugf. This also works concurrently, etc. Have you tried that? Why didn't it work?

Closing this, but feel free to reopen if there's something I'm missing.

**Comment by mvo5** [comment]
Use "-vv" instead of showOutput and move the current -vv to -vvv .

**Comment by sergiocazzolato** [comment]
@mvo5 I already updated this change according to the review done in Frankfurt.
Thanks

File: spread/client.go [hunk 1]
```diff
@@ -468,13 +476,21 @@ func (c *Client) runPart(script string, dir string, env *Environment, mode outpu
...
 
 	if e, ok := err.(*ssh.ExitError); ok && e.ExitStatus() == 213 {
+		lines := bytes.Split(bytes.TrimSpace(stdout.Bytes()), []byte{'\n'})
+		m := commandExp.FindSubmatch(lines[len(lines)-1])
+		if len(m) > 0 && string(m[1]) == "ERROR" {
+			return nil, fmt.Errorf("%s", m[2])
+		}
+	}
+
+	if e, ok := err.(*ssh.ExitError); ok && e.ExitStatus() == 214 {
```
Comment (mvo5) [inline spread/client.go:486]: I think this needs a comment why we now have two different exit codes and why the old detection no longer works. It's also slightly dangerous as now any exit code 214 in -vv mode will trigger a reboot. I wonder if there is a way to make this nicer. I will meditate a bit.

File: spread/project.go [hunk 2]
```diff
@@ -974,6 +974,7 @@ func (p *Project) Jobs(options *Options) ([]*Job, error) {
 
 	all := jobs
 	jobs = make([]*Job, 0, len(all))
+
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply spread/project.go:977]: fixed

File: spread/runner.go [hunk 3]
```diff
@@ -473,20 +485,28 @@ func (r *Runner) run(client *Client, job *Job, verb string, context interface{},
...
 		start = start.Add(1)
 		printft(start, startTime|endTime|startFold|endFold, "Error %s %s (%s) : %v", verb, contextStr, server.Label(), err)
 		if debug != "" {
-			start = time.Now()
-			output, err := client.Trace(debug, dir, job.Environment)
-			if err != nil {
-				printft(start, startTime|endTime|startFold|endFold, "Error debugging %s (%s) : %v", contextStr, server.Label(), err)
-			} else if len(output) > 0 {
-				printft(start, startTime|endTime|startFold|endFold, "Debug output for %s (%s) : %v", contextStr, server.Label(), outputErr(output, nil))
+			if ! r.options.ShowOutput {
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply spread/runner.go:502]: done

File: tests/output/task.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    if [ ! -f .spread-reuse.yaml ]; then
+        touch /run/spread-reuse.yaml
+        ln -s /run/spread-reuse.yaml .spread-reuse.yaml
+    fi
+
+restore: |
+    rm -f exec1.out exec2.out exec3.out
+
+execute: |
+    # Check the task output is shown when -show-output is not used
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/output/task.yaml:13]: done

File: spread/client.go (hunk 1 above)
Comment (sergiocazzolato) [reply spread/client.go:486]: I'll need to reproduce the error because I don't remember exactly the details

Comment (sergiocazzolato) [reply spread/client.go]: I already removed the new exit code and left 213 where the -vv is handled

**Comment by sergiocazzolato** [comment]
@mvo5 all the comments have been addresses. Thanks for the review

**Comment by mvo5** [comment]
Discussed with Gustavo - streaming can be simulated with a very short warn-timeout (configurable already) plus -single-worker. So this needs a stronger use-case to be considered. The trouble with this PR is that it breaks buffering and that breaks the massiv parallel model that spread is using.

**Pull request closed without merging**